File: 35 - Merge Overlapping Time Intervals/solution.py
def merge_intervals(intervals):
    if not intervals:
        return []
    
    intervals.sort(key=lambda x: x[0])

    merged = [intervals[0]]

    for i in range(1, len(intervals)):
        last = merged[-1]
        current = intervals[i]

        if current[0] <= last[1]:
            # Overlap detected, merge them
            last[1] = max(last[1], current[1])
        else:
            # No overlap, add current interval
            merged.append(current)

    return merged

# Test Cases
print(merge_intervals([[1, 3], [2, 6], [8, 10], [15, 18]])) 

File: 35 - Merge Overlapping Time Intervals/solution.js
function mergeIntervals(intervals) {
  if (intervals.length === 0) return [];

  intervals.sort((a, b) => a[0] - b[0]);

  const merged = [intervals[0]];

  for (let i = 1; i < intervals.length; i++) {
    const last = merged[merged.length - 1];
    const current = intervals[i];

    if (current[0] <= last[1]) {
      // overlapping, merge current and last
      last[1] = Math.max(current[1], last[1]);
    } else {
      // not overlapping, push last and start a new one
      merged.push(current);
    }
  }

  return merged;
}

// Test Cases
console.log(
  mergeIntervals([
    [1, 3],
    [2, 6],
    [8, 10],
    [15, 18],
  ])
);
